strawberry/types/type_resolver.py

```diff
@@ -138,23 +122,25 @@ class if one is not set by either using an explicit strawberry.field(name=...) o
             # Note: We do this here rather in the `Strawberry.type` setter
             # function because at that point we don't have a link to the object
             # type that the field as attached to.
-            if isinstance(field.type_annotation, StrawberryAnnotation):
-                type_annotation = field.type_annotation
-                type_annotation.annotation = _resolve_specialized_type_var(
-                    cls,
-                    cast(type, type_annotation.annotation),
-                )
-                if type_annotation.namespace is None:
-                    type_annotation.set_namespace_from_field(field)
+            if (
+                isinstance(field.type_annotation, StrawberryAnnotation)
+                and field.type_annotation.namespace is None
+            ):
+                field.type_annotation.set_namespace_from_field(field)
+
+            # If the field origin is not cls, copy it and set cls on it.
+            # This is important for us to be able to resolve specialized
+            # generic when generating the schema.
+            if field.origin != cls:
+                field = copy.copy(field)  # noqa: PLW2901
+                field.origin = cls
```

Comment:
do we need to do this all the time? shouldn't we check if it is generic?

Reply:
I wanted to do that, but it all falls back to the same "forward ref" issues. There's no way to be sure that this is generic when using forward refs at this point, because we can't properly resolve the type until schema creation.